working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+
+```dart
+// dart:concurrent
+
+abstract class Thread implements Shareable {
+  /// Runs the given function in a new thread.
+  ///
+  /// The function is run in a shared isolate, meaning that 
+  /// it will not have access to the non-`@shared` state.
+  ///
+  /// The function will be run in a `Zone` which uses the
+  /// spawned thread as an executor for all callbacks: this
+  /// means the thread will remain alive as long as there is 
+  /// a callback referencing it.
+  external static Thread start<T>(FutureOr<T> Function() main); 
+  
+  external Thread get current;
+    
+  Future<void> join();
+  
+  void interrupt();
+    
+  external set priority(ThreadPriority value);
+  ThreadPriority get priority;
+}
+
+abstract class ThreadPool implements Shareable {
+  external factory ThreadPool({required int concurrency});
+  
+  /// Run the given function on this thread pool.
+  /// 
+  /// The [task] is executed in a shared isolate of the current
+  /// isolate group. 
+  ///
+  /// The function will be run in a `Zone` which uses this
+  /// pool as an executor for all callbacks: this
+  /// means the pool will remain alive as long as there is 
+  /// a callback referencing it.
+  void postTask(void Function() task);
+}
+```
+
+We might also want to provide a way to synchronously execute code in a specific isolate on the _current_ thread:
+
+```dart
+class Isolate {
+    T runSync<T>(T Function() cb);
+}
+```
+
+##### Asynchronous code and threads
+
+Consider the following example:
+
+``` dart
+Thread.start(() async {
+  var v = await foo();
+  var u = await bar();
+});
+```
+
+Connecting execution / scheduling behavior to `Zone` allows us to give a clear semantics to this code: this code will run on a specific (newly spawned) thread and will not change threads between suspending and resumptions.  
+
+#### Atomic operations
+
+`Atomic<T>` is a wrapper around a value of type `T` which can be updated atomically. It can only be used with true reference types - an attempt to create an `AtomicRef<int>`, `AtomicRef<double>` , `AtomicRef<(T1, ..., Tn)>` will throw.  
+
+```dart
+// dart:concurrent
+
+class AtomicRef<T extends Shareable> implements Shareable {
+  /// Atomically updates the current value to [desired].
+  ///
+  /// The store has release memory order semantics.
+  void store(T desired);
+
+  /// Atomically reads the current value.
+  ///
+  /// The load has acquire memory order semantics.
+  T load();
+  
+  /// Atomically compares whether the current value is identical to
+  /// [expected] and if it is sets it to [desired] and returns 
+  /// `(true, expected)`.
+  ///
+  /// Otherwise the value is not changed and `(false, currentValue)` is
+  /// returned.
+  (bool, T) compareAndSwap(T expected, T desired);
```

Comment:
I realize that "load", "store", and "compare and swap" is established terminology, particularly in the compiler engineering space, but I think we can pick something that's more idiomatic Dart. Because the type declaration - `AtomicRef... implements Shareable` - already strongly implies atomicity and shared access there's no need to reinforce that at the API level. This is similar to how `List` and `Stream` imply what happens when you call their respective `add` methods. We didn't call those methods `List.append` and `Stream.sink`, for example, to underscore that lists add values at the end, and streams fire and forget.

Here's a more darty variation of the API:

```
class AtomicRef<T extends Shareable> implements Shareable {
  T get value;
  set value(T newValue);
  (bool, T) replace(T currentValue, T newValue);
}
```

Reply:
> I realize that "load", "store", and "compare and swap" is established terminology, particularly in the compiler engineering space, but I think we can pick something that's more idiomatic Dart.

I think replacing `load` and `store` with a getter / setter pair _might_ be okay, but changing `compareAndSwap` to `replace` is definitely not okay. This has nothing to do with compiler construction. It's an established [terminology](https://en.wikipedia.org/wiki/Compare-and-swap). If somebody says "CAS" you either know what that means or you better not try to write any code with atomics.

My motivation for using `load` and `store` instead of getter and setter was to align with extension methods (which use `atomicLoad` and `atomicStore` naming) and to indicate that this does something more complicated then just storing a value (e.g. there is a barrier to ensure release semantics).

Reply:
> If somebody says "CAS" you either know what that means or you better not try to write any code with atomics.

Ha-ha! That's a solid point.